Refactor marketplace code

MikelAlejoBR opened the pull request:
Essentially what this PR does is:

1. Moves all the marketplace related code to the "marketplace" package.
3. Removes the "append the marketplace token to authentication's extra field" code from the DAO layer, and puts it in the handlers.
4. Makes it easier to use because now you only need to call `marketplace.SetMarketplaceTokenAuthExtraField(tenantId, auth)` and it takes care of everything for you. Prior to this we had to be initializing structs with the `tenantId` here and there which felt like spaghetti code.
5. Extracts all the interfaces to an "interfaces.go" file at the package level, so that it is easier to see what is the package about.
6. Logs the error messages as detailed as possible, and returns a generic error message to the user. This is because usually the user can't do much about a "could not parse int from str" message.

MikelAlejoBR committed:
refactor: move all the marketplace related code to its own package

Instead of having everything scattered around, it makes sense having
everything in one place, to reduce coupling code to specific
implementations.

MikelAlejoBR committed:
refactor: place the marketplace token management in the handlers

Since we don't need to have the token persisted anywhere but Redis, we
don't need to have the operations at the DAO level. Plus, it decouples
the token management from the data access itself, and makes it easier to
not forget about when dealing with authentications.

MikelAlejoBR committed:
refactor: extract the interfaces to a single file

By having them in the same file it gives a overview of what the package
is about.

MikelAlejoBR committed:
refactor: improve logging and error messages

Until now we usually logged the error that the functions outputted, and
we ended up returning that same error to the user. The problem is that
generally those errors are "unrecoverable" from the user's standpoint:
they can't do much about it.

Therefore I think it makes sense to return a generic error message along
with an "internal server error" response to let them know that it wasn't
their fault, and that they should report the error whenever it is possible.

On the other hand, in our case we should be logging the error with as
much context as possible, and that's why I changed the log messages for
the marketplace code.

lindgrenj6 commented on marketplace/interfaces.go:
```diff
@@ -0,0 +1,21 @@
+// tests.
+type HttpClient interface {
+	Do(req *http.Request) (*http.Response, error)
+}
```
since we're refactoring this - would you be ok with using a request with a context rather than the custom http client? it removes the need for the interface + getter methods.

http.NewRequest -> http.NewRequestWithContext

I did it for rbac here: https://github.com/RedHatInsights/sources-api-go/blob/97c2a6316d83f7ee2363c7a911fca6d2e72f3e30/middleware/authorization.go#L120-L121

--- 
I totally get it if not - just a thought.

MikelAlejoBR replied:
It seems there is a way to "mock" the HTTP server with `httptest`: https://stackoverflow.com/a/51227660. Let me try that to see if I can remove the interface...

MikelAlejoBR replied:
Done!

lindgrenj6 reviewed (approved):
This looks good - I just had a thought about the custom http interface.

MikelAlejoBR committed:
refactor: remove the HttpClient interface

It is not needed since the tests can be run with the "httptest" library
instead, which allows us to do what we want.

MikelAlejoBR committed:
refactor: tidy up the tests a little bit

The "err == nil" is not necessary since we specifically check for a
particular error message below, so the "err == nil" case would also
be covered.

lindgrenj6 commented on marketplace/marketplace_token_provider.go:
```diff
@@ -50,21 +31,27 @@ func (mtp *MarketplaceTokenProvider) RequestToken() (*BearerToken, error) {
 	request.Header.Add("Content-Type", "application/x-www-form-urlencoded")
 
-	client := GetHttpClient()
+	client := http.Client{Timeout: 3 * time.Second}
```
I guess maybe I wasn't clear - rather than having a custom HTTP client I was thinking of just using `http.DefaultClient.Do()`, but then up where `http.NewRequest` is called we can just use `http.NewRequestWithContext`. 

like so: 
```diff
diff --git a/marketplace/marketplace_token_provider.go b/marketplace/marketplace_token_provider.go
index 4bb19c1..d04bd1e 100644
--- a/marketplace/marketplace_token_provider.go
+++ b/marketplace/marketplace_token_provider.go
@@ -1,6 +1,7 @@
 package marketplace
 
 import (
+	"context"
 	"fmt"
 	"net/http"
 	"net/url"
@@ -43,7 +44,12 @@ func (mtp *MarketplaceTokenProvider) RequestToken() (*BearerToken, error) {
 	data.Set("apikey", *mtp.ApiKey)
 	data.Set("grant_type", "urn:ibm:params:oauth:grant-type:apikey")
 
-	request, err := http.NewRequest(
+	// set up a context to control the timeout of the request
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	request, err := http.NewRequestWithContext(
+		ctx,
 		"POST",
 		config.Get().MarketplaceHost+"/api-security/om-auth/cloud/token",
 		strings.NewReader(data.Encode()),
@@ -57,8 +63,7 @@ func (mtp *MarketplaceTokenProvider) RequestToken() (*BearerToken, error) {
 	request.Header.Add("Accept", "application/json")
 	request.Header.Add("Content-Type", "application/x-www-form-urlencoded")
 
-	client := GetHttpClient()
-	response, err := client.Do(request)
+	response, err := http.DefaultClient.Do(request)
 	if err != nil {
 		return nil, fmt.Errorf("could not perform the request to the marketplace: %s", err)
 	}
```

MikelAlejoBR replied:
Sure! Done.
